### PR TITLE
Fix line endings in the Fem workbench.

### PR DESCRIPTION
**Pull request by ezzieyguywuf**
Per (this discussion)[https://forum.freecadweb.org/viewtopic.php?f=17&t=41117&p=351671#p351670], this PR standardizes line-endings in the Fem workbench.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [x] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- n/a Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by ezzieyguywuf** [comment]
Sure thing! Can you add these changes to the pr or do I need to push a new
commit myself? (I’m currently away from my computer)

**Comment by berndhahnebach** [comment]
No hurry. It would be great if you could add them if you are back on your computer, but as said no hurry. FreeCAD motto is: "It's done when it's done."

**Comment by ezzieyguywuf** [comment]
Done - I used the `magic` python module to check every file in the Fem workbench - anything that returned with a mime-type of "text", I added it's extension to our `.gitattributes` file.

I should note that the following files are text files, but don't have file extensions. As such, I'm not entirely sure how git will handle them:

- src/Mod/Fem/femtest/data/elmer/ELMERSOLVER_STARTINFO
- src/Mod/Fem/femtest/data/ccx/Flow1D_thermomech_expected_values
- src/Mod/Fem/femtest/data/ccx/cube_frequency_expected_values
- src/Mod/Fem/femtest/data/ccx/cube_static_expected_values
- src/Mod/Fem/femtest/data/ccx/spine_thermomech_expected_values

I should also note that the file `src/Mod/Fem/femtest/data/ccx/Flow1D_thermomech.dat` is empty, and is classified as an "inode" type (not sure what that means).

Finally, for posterity, here is the python script I used to determine which file extensions to add `.gitattributes`

```py
import os
import magic
import pprint

mymap = {}
for dirpath,dirnames,filenames in os.walk('src/Mod/Fem'):
    for fname in filenames:
        check_name = os.path.join(dirpath, fname)
        val = magic.from_file(check_name, mime=True)
        left, right = val.split('/')
        root, extension = os.path.splitext(fname)
        if not left in mymap.keys():
            mymap[left] = {}
        if not extension in mymap[left].keys():
            mymap[left][extension] = []
        mymap[left][extension].append(check_name)

pprint.pprint(mymap)
```

**Comment by ezzieyguywuf** [comment]
I believe the Travis failure was due to compile time. I'm going to close and re-open the PR to re-trigger the travis, hopefully the pre-existing cache the first build created will reduce the build time down to an allowable level.

File: .gitattributes [hunk 1]
```diff
@@ -7,6 +7,23 @@
...
 src/Mod/Fem/*.csv text
+src/Mod/Fem/*.py text
+src/Mod/Fem/*.xml text
+src/Mod/Fem/*.txt text
+src/Mod/Fem/*.md text
+src/Mod/Fem/*.ui text
+src/Mod/Fem/*.qrc text
+src/Mod/Fem/*.bat text
+src/Mod/Fem/*.dat text
+src/Mod/Fem/*.ts text
```
Comment (berndhahnebach) [inline .gitattributes]: these are translation files. Since I am not sure  if this could cause trouble we should comment  it. Furthermore we should only use a file ending here if we are sure all files of this ending have 100 % unix line endings.

File: .gitattributes [hunk 2]
```diff
@@ -7,6 +7,23 @@
...
+src/Mod/Fem/*.unv text
+src/Mod/Fem/*.yml text
+src/Mod/Fem/*.vtk text
+src/Mod/Fem/*.inp text
+src/Mod/Fem/*.z88 text
+src/Mod/Fem/*.sif text
+src/Mod/Fem/*.geo text
+src/Mod/Fem/*.frd text
```
Comment (berndhahnebach) [inline .gitattributes]: why the comment   "This really should not be needed..."  What does it mean?

File: .gitattributes (hunk 1 above)
Comment (ezzieyguywuf) [reply .gitattributes]: Per my investigations, at this point (i.e. after commit ab2f02), all files in the Fem sub-directory structure have unix-style line-endings. By including these file extensions in the `.gitattributes` file, we ensure that these file remain as such moving forward.

Finally, based on my mime-type studies using the python script provided in my previous comment, all four of the file-extensions you list here are, in fact, simple text-files.

**Comment by ezzieyguywuf** [comment]
I wanted to provide an example of how to explicitly tell git "always check
out this file with CRLF-style (i.e. windows-style) line endings". This is
in case, in the future, someone discovers an issue with allowing git to
convert the line-endings of some particular file.

By "This really should net be needed..." I mean that there is no real
reason to _force_ git to _always_ checkout a file with CRLF line endings.

If the proper "core.autocrlf" setting is used, then windows-users will have
CRLF endings upon checkout, and linux-users will have unix-style line
endings. Upon checking, everything gets unified to unix-style line endings.

However, as I'm not 100% familiar with all the intricasies of how windows
works, I imagine there may be a situation where _checking-in a
windows-style line-ending file and then checking it back out results in
problems_. In this case, and upon careful consideration, this option should
be used.

File: .gitattributes (hunk 1 above)
Comment (berndhahnebach) [reply .gitattributes]: great this is very important!

My comment was only for the *.ts files. The others are fine. I just have never ever done anything in the regard of the *.ts files. Thus I do not know if we could run intot trouble here.

**Comment by berndhahnebach** [comment]
Thanks for the explanation. I got the point.

**Comment by berndhahnebach** [comment]
https://forum.freecadweb.org/viewtopic.php?f=17&t=41117
just to have the link here too

**Comment by berndhahnebach** [comment]
I have been reading about gitattributes a bit ... 
Are you 110 % sure only files in Fem will be touched (normalised)?

**Comment by berndhahnebach** [comment]
should we add svg to the binaries to be sure?

**Comment by ezzieyguywuf** [comment]
> I have been reading about gitattributes a bit ...
> Are you 110 % sure only files in Fem will be touched (normalised)?

What part of the gitattributes documentation has you worried? If you examine [my gitattribute changes](https://github.com/FreeCAD/FreeCAD/pull/2752/files#diff-fc723d30b02a4cca7a534518111c1a66R7), you'll notice that I fully quality the path to the Fem workbench, rather than simply wild-carding all files with these file extensions.

When I check [the documentation](https://git-scm.com/docs/gitattributes), it states that:

> When the pattern matches the path in question, the attributes listed on the line are given to the path.

In other words, no other files should match these patterns, since no other files are located in `src/Mod/Fem` except for Fem  files.

But am I "110% sure"....no. Because I have not tested this exhaustively.

> should we add svg to the binaries to be sure?

Yes, I think I just missed this when I was reviewing the output of my python script. I've already pushed a commit to add this.

**Comment by berndhahnebach** [comment]
try `git add --renormalize .` on FreeCAD master and on your branch. It will change half of the repo. I have been reading more about gitattributes. Here is what I would come up with. https://github.com/berndhahnebach/FreeCAD_bhb/commits/femfileendings
What do you think of this?

Still I need to test with explizit commits what happens in different circumstanzes.

**Comment by ezzieyguywuf** [comment]
>  What do you think of this?

Your approach is correct, we should be the Fem-specific `gitattributes` in `src/Mod/Fem` as you have done.

> Still we need to test with explizit commits what happens in different circumstanzes.

I don't know what you mean by this.

**Comment by berndhahnebach** [comment]
We need to play a bit, change some code in Fem and make a line ending to windows commit see what happens. Change something outside Fem commit see what happens. ETC. All this on a github repo. Just try some real examples to be sure we are on the right track before this is merged.

bernd

**Comment by berndhahnebach** [comment]
should we close this one and I will open one with my changes and we continue the disscussion over there. Bernd

**Comment by ezzieyguywuf** [comment]
👍🏽 Let’s do it!

**Comment by ezzieyguywuf** [comment]
Closing in favour of bernd‘s branch

**Comment by berndhahnebach** [comment]
https://github.com/FreeCAD/FreeCAD/pull/2757
